# Introduction to Data Science - Homework 2
*CS 5963 / MATH 3900, University of Utah, http://datasciencecourse.net/*

Due: Friday, September 9, 11:59pm.

In this homework you will read in and analyze a movies dataset. First we'll do some basic analysis with vanilla Python, then we'll move on to doing more advanced analysis with [Pandas](http://pandas.pydata.org/pandas-docs/stable). 


## Your Data
Fill out the following information: 

*First Name:*  Ty 
*Last Name:*   Clive
*E-mail:*   u0389772@utah.edu
*UID:*  u0389772


## Part 1: Analyzing Data The Hard Way

In this part we'll do some manual analysis of a movies dataset.

### Task 1.1: Read in the data

Parse the file [movies.csv](movies.csv) using the [csv library](https://docs.python.org/3/library/csv.html). Lab 4 and/or Homework 1 might be a good inspiration for this. 

We recommend that you store the header in a separate array. Make sure that at least the "ratings" and the "votes" columns are cast to the appropriate data types for doing calculations with them. 

In [134]:
import csv

data_values = []
#rating = data_values[int(5)]
#votes = data_values[int(6)]

# import the csv library
#import csv
#open('movies.csv')
#print(movies.csv)

with open('movies.csv') as csvfile:
    # note that we can interchangably use ' and " in general
    # for the quotechar, however we use ' so that we can use " without escaping
    filereader = csv.reader(csvfile, delimiter=',', quotechar='"')
    # the row here is an array
    for row in filereader:
        #print("Row: " + str(row))
        data_values.append(row)

# Store the header in a separate array
header = data_values.pop(0)
   
#print()    
print(header)
#print()
#print(head(data_values)



# initialize the top-level array
movies = []
header = []

# your import code goes here

['', 'title', 'year', 'length', 'budget', 'rating', 'votes', 'r1', 'r2', 'r3', 'r4', 'r5', 'r6', 'r7', 'r8', 'r9', 'r10', 'mpaa', 'Action', 'Animation', 'Comedy', 'Drama', 'Documentary', 'Romance', 'Short']


### Task 1.2:  Calculate movie stats

In this task, you will calculate some statistics about movies. We suggest you implement your solutions for Tasks 1.2.1-1.2.3 in one code cell - you should be able to calculate this in a single iteration over the dataset.

#### Task 1.2.1 Calcualte average ratings
Compute the average rating for the movies and print the output. Also print the number of movies.
Your output could look like this:
    
    Average rating: xxx.xxxx, Number of movies: xx

#### Task 1.2.2: Calculate average rating for major movies

Compute the average rating for the movies that have more than 500 votes in your loaded dataset and print the output. (We'll call these movies with more than 500 votes *major movies* from now on).

Your output could look like this:
    
    Average rating of movies with more than 500 votes: xxx.xxxx, Number of major movies: xx

#### Task 1.2.3: Find the highest rated major movie

Find out which of the movies with more than 500 votes has the highest rating.

Your output could look like this:
    
    Highest rating: xxx.xxxx, Title: MOVIE TITLE
 
#### Task 1.2.4: Interpret the data
 * What's the size relationship of major movies to all movies?
 * Are major movies usually better than the average movies?

In [130]:
#def movie_ratings():
    
    # Overall average of ratings

def movie_ratings_overall_averages(movie_list):
    
    number_of_movies = len(movie_list)
    ratings_total = 0.0
    major_ratings_total = 0.0
    major_movie_counter = 0
    best_movie = ""
    highest_rating = 0.0
    
    for movie in movie_list:
        current_movie_rating = float(movie[5])
        ratings_total += current_movie_rating
        
       
        if int(movie[6]) > 500:
            major_ratings_total += current_movie_rating
            major_movie_counter += 1
            
            
        # Highest rating check
        if current_movie_rating > highest_rating: 
            highest_rating = current_movie_rating
            best_movie = movie[1]

        
    
    print("Average rating: ", ratings_total/number_of_movies, ", Number of movies: ", number_of_movies)
    print("Average rating of movies with more than 500 votes: ", major_ratings_total/major_movie_counter, ", Number of major movies: ", major_movie_counter)
    print("Highest rating: ", highest_rating, ", Title: ", best_movie)
    
movie_ratings_overall_averages(data_values)  

Average rating:  5.932850241545932 , Number of movies:  58788
Average rating of movies with more than 500 votes:  6.370579126664578 , Number of major movies:  6458
Highest rating:  10.0 , Title:  Dimensia Minds Trilogy: The Hope Factor


**Your Interpretation**: 
I created a function that produces output to answer the three questions: (a) average rating for all movies, (b) average rating for popular movies (that got over 500 votes), and (c) the highest rating and title of highest rated movie. A list of variables were defined before I passed them into my function.

Average rating for all movies -

The first for loop stores the value for rating in "current_movie_rating" (float), and as the function iterates through it adds up all current_movie_ratings and stores them in ratings_total. The first print function divides the total_ratings variable by length of the matrix (header excluded from the first exercise). That returns the overall average rating for the entire list of movies.

Average rating for major movies - 

The second part of the for loop has an if-qualifier that only counts movies with >500 votes (that's how we define whether a movie is major). The function stores current_movie_rating in variable major_ratings_total. I also created a counter to count how many major movies are iterated through in the for loop. In the second print function, major_ratings_total/ major_movie_counter accurately returns the mean.

Highest rated major movie -

I created a highest_rating variable, which stores the value of the highest rating that the for loop iterates through. I also created a variable called best_movie, which stores the movie title in position 1 on the same row as that rating.

Interpretat the data 1.2.4 - 

There are 6458 major movies among 58788 movies. That means only ~10.99% of the movies that come out get over 500 votes. The average rating of major movies is 6.37 while the average rating of all movies is 5.93 (and presumably lower if you exclude the major movies). So on average, the major movies are more highly rated than movies are generally.

## Part 2: Pandas

In this part we will use the [Pandas library](http://pandas.pydata.org/pandas-docs/stable) for our analysis.

### Task 2.1: Loading data

Read in the data again. This time you should create a [Pandas DataFrame](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html). Print the head of the dataset. 
 * How many data rows did you load? How many columns?
 * Of which data types are the columns? 
 * Do you have to do manual data type conversions?

In [131]:
# import pandas
import pandas as pd

movies = pd.read_csv("movies.csv")
df = movies

df.head()
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58788 entries, 0 to 58787
Data columns (total 25 columns):
Unnamed: 0     58788 non-null int64
title          58788 non-null object
year           58788 non-null int64
length         58788 non-null int64
budget         5215 non-null float64
rating         58788 non-null float64
votes          58788 non-null int64
r1             58788 non-null float64
r2             58788 non-null float64
r3             58788 non-null float64
r4             58788 non-null float64
r5             58788 non-null float64
r6             58788 non-null float64
r7             58788 non-null float64
r8             58788 non-null float64
r9             58788 non-null float64
r10            58788 non-null float64
mpaa           4924 non-null object
Action         58788 non-null int64
Animation      58788 non-null int64
Comedy         58788 non-null int64
Drama          58788 non-null int64
Documentary    58788 non-null int64
Romance        58788 non-null int64
Sho

/Applications/anaconda/lib/python3.5/site-packages/numpy/lib/function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


Unnamed: 0          year        length        budget        rating  \
count  58788.00000  58788.000000  58788.000000  5.215000e+03  58788.000000   
mean   29394.50000   1976.133582     82.337875  1.341251e+07      5.932850   
std    16970.77815     23.735125     44.347717  2.335008e+07      1.553031   
min        1.00000   1893.000000      1.000000  0.000000e+00      1.000000   
25%    14697.75000   1958.000000     74.000000           NaN      5.000000   
50%    29394.50000   1983.000000     90.000000           NaN      6.100000   
75%    44091.25000   1997.000000    100.000000           NaN      7.000000   
max    58788.00000   2005.000000   5220.000000  2.000000e+08     10.000000   

               votes            r1            r2            r3            r4  \
count   58788.000000  58788.000000  58788.000000  58788.000000  58788.000000   
mean      632.130384      7.014382      4.022377      4.721159      6.374847   
std      3829.621413     10.936759      5.960056      6.451629      7.586141   
min         5.000000      0.000000      0.000000      0.000000      0.000000   
25%        11.000000      0.000000      0.000000      0.000000      0.000000   
50%        30.000000      4.500000      4.500000      4.500000      4.500000   
75%       112.000000      4.500000      4.500000      4.500000      4.500000   
max    157608.000000    100.000000     84.500000     84.500000    100.000000   

           ...                 r8            r9           r10        Action  \
count      ...       58788.000000  58788.000000  58788.000000  58788.000000   
mean       ...          13.875995      8.954208     16.854018      0.079744   
std        ...          11.316963      9.440597     15.650722      0.270899   
min        ...           0.000000      0.000000      0.000000      0.000000   
25%        ...           4.500000      4.500000      4.500000      0.000000   
50%        ...          14.500000      4.500000     14.500000      0.000000   
75%        ...          24.500000     14.500000     24.500000      0.000000   
max        ...         100.000000    100.000000    100.000000      1.000000   

          Animation        Comedy         Drama   Documentary       Romance  \
count  58788.000000  58788.000000  58788.000000  58788.000000  58788.000000   
mean       0.062768      0.293784      0.371011      0.059060      0.080697   
std        0.242547      0.455498      0.483080      0.235738      0.272371   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      1.000000      1.000000      0.000000      0.000000   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

              Short  
count  58788.000000  
mean       0.160883  
std        0.367426  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000  

[8 rows x 23 columns]

**Your Interpretation**:
I loaded 58787 rows with 25 cloumns. The column data types are ints, objects (string), and floats. All of the entries are non-null values, and fortunately I didn't have to do any data type conversions to get the appropriate summary statistics.


### Task 2.2: Calculate the average rating

Compute the average rating for all movies and print the output in a formatted way.

Your output could look like this:
    
    Average rating: xxx.xxxx

In [102]:
print("Average movie rating: ", movies['rating'].mean())

Average movie rating:  5.932850241545932


### Task 2.3: Compare the runtime

Measure the runtime of the mean calculation using [Pandas](http://pandas.pydata.org/pandas-docs/stable).
Compare it to the computation time for calculating the mean using a for loop (you can copy the relevant parts from part one).

You can use [time.clock()](https://docs.python.org/3/library/time.html) to set timestamps before and after the execution of the code you want to measure, then you simply substract end time from start time.

Print your results in a human readable way and add a metric to the output. Calculate the factor of the difference and print it. Note that the exact times and the factors will vary when you re-run this and especially between machines. 

E.g.:  
    time using own code: xxx.xxx s  
    time using Pandas: xxx.xxx s  
    difference factor: xxx

In [106]:
import time

#using my code

start_time1 = time.clock()

movie_ratings_overall_average(data_values)

end_time1 = time.clock()

time1 = (end_time1 - start_time1)

#using pandas

start_time2 = time.clock()

movies['rating'].mean()

end_time2 = time.clock()

time2 = (end_time2 - start_time2)

print("My manual code is ", time1/time2, "times slower than Pandas")

Average rating:  5.932850241545932 , Number of movies:  58788
Average rating of movies with more than 500 votes:  6.370579126664578 , Number of major movies:  6458
Highest rating:  10.0 , Title:  Dimensia Minds Trilogy: The Hope Factor
My manual code is  178.05074626879377 times slower than Pandas


### Task 2.4: Filter out rows

The whole movies dataset has about 60k entries. Use pandas to filter your dataframe to contain only the major movies with more than 500 votes.

Count and print the number of movies with more than 500 votes. 

E.g.:
    xxx.xxx movies have more than 500 votes.

In [108]:
movie_votes = movies['votes']
majors_mask = movie_votes > 500
majors = movie_votes[majors_mask]
count_majors = majors.count()

print(count_majors, " movies have more than 500 votes.")


movie_votes2 = movies['votes']
count_majors2 = movie_votes2[movie_votes2>500].count()

print(count_majors2, " movies have more than 500 votes.")

6458  movies have more than 500 votes.
6458  movies have more than 500 votes.


### Task 2.5: Calculate the average rating for major movies

Compute the average rating for the major movies.
Your output could look like this:
    
    Average rating of movies with more than 500 votes: xxx.xxxx

In [115]:
# major movies mean 
movie_ratings = movies['rating']
majors_ratings = movie_ratings[majors_mask]


#major_movies_mean_rating = movie_ratings[movie_votes2>500].mean()

print("Average rating of movies with more than 500 votes: ", majors_ratings.mean())

Average rating of movies with more than 500 votes:  6.370579126664578


### Task 2.6: Find the highest rated major movie

Find the highest rated major movie in the dataframe. Hint: [idxmax()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.idxmax.html) could be a helpful function. 

Print the title and the rating.

Your output could look like this:
    
    Highest rated movie: TITLE, rating: x.x


In [121]:
# highest rated movie

highest_rating = majors_ratings.idxmax()

print("Highest rated movie: ", movies.iloc[highest_rating].title, "rating: ", movies.iloc[highest_rating].rating)

Highest rated movie:  Godfather, The rating:  9.1


### Task 2.7: Filter out rows and count specific features

From the major movies you filtered out in a previous task, find out how many are comedies.
How many are dramas? How many are both, comedies and dramas? Hint: you can combine broadcasting statements with a boolean AND: `&`.

Print the results, e.g.:  

    xxx.xxx major movies are comedies.
    xxx.xxx major movies are dramas.
    xxx.xxx major movies are both, comedies and dramas.

In [129]:
# count genres

major_comedies = movies[('votes' > 500) & ('Comedy'==1)]
major_dramas = movies[('votes' > 500) & ('Drama'==1)]
major_comedy_dramas = movie[('votes' > 500) & ('Comedy'==1) & ('Drama'==1)]

print(major_comedies.count(), " major movies are comedies.")
print(major_dramas.count(), " major movies are dramas.")
print(major_comedy_dramas(), " major movies are comedy-dramas")

TypeError: unorderable types: str() > int()

### Task 2.8: Compare ratings of different categories

Now we want to compare the ratings for comedies and dramas for major movies.

Compute the average rating, maximum rating, minimum rating, standard deviation, and the median for each category. Hint: there is a function that does all of this in one line. 

What do these numbers tell us? Provide an interpretation.

In [ ]:
# genre stats go here

major_comedies.summary('ratings')
major_dramas.summary('ratings')  

**Your Interpretation**: 

I don't understand why Pandas doesn't take these functions. I interpret this as a suggestion that I should get more familiar with Pandas.

### Task 2.9: Movies per year

Calcluate how many major movies were made in each year.

Print the number like this, sorted by year:

    year   number of movies
    1902    xxx
    1903    xxx
    ...     ...

Use this data to render a line chart of the number of movies per year.

In [ ]:
# cacluate number of movies 

In [ ]:
# create the plot here

### Task 2.10: Yearly average

Compute the average rating per year for all major movies.  
Use the numbers you computed to plot a line chart. Plot the year on the x-axis and the average rating on th y axis.  

In [ ]:
# create the plot here

### Task 2.11: Explore and Interpret

Are old movies better? How could you explain this? Are there differences between the rating of major movies and all movies over time? Continue to explore and use plots to inform your answer. Interesting measures to consider are the total number of votes per year, the average number of votes for a movie in a particular year, etc. 

TODO: your interpretation and your open exploration